### PR TITLE
Include support for GQL custom scalar types

### PR DESCRIPTION
**Pull request by ahussain3**
In the GraphQL schema, it is possible to specify (CustomScalarTypes)[https://www.apollographql.com/docs/graphql-tools/scalars.html].

In order to specify a Custom Scalar Type, you need to define an object which implements the following interface:
```
serialize()
parse_literal()
parse_value()
```
These methods allow GraphQL to translate between the serialized and deserialized versions of the custom scalar object.

However, when the GQL schema is passed around it includes only the names of the custom scalar types, and not the serialization/deserialization methods. After this PR is merged, you should be able to define your own classes on the client, and by giving them a parse_value() method you can use them as custom scalar types.

To configure the gql client to use your custom scalar types, you should do:
```
from gql import gql, Client
custom_types = {
    'DateTime': MyDateTimeClass
}

client = Client(schema=schema, transport=transport, custom_types=custom_types)
response = client.execute(query)
```

### CODE REVIEW

File: gql/response_parser.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+from typing import Any, Dict, Callable, Optional, List
+
+from graphql.type.schema import GraphQLSchema
+from graphql.type.definition import GraphQLObjectType, GraphQLField, GraphQLScalarType
+
+
+class ResponseParser(object):
```
Comment (lincolnq) [inline gql/response_parser.py]: needs a different name - something involving type adapters or whatever

File: gql/transport/requests.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ def __init__(self, url, auth=None, use_json=False, timeout=None, **kwargs):
         self.default_timeout = timeout
         self.use_json = use_json
 
-    def execute(self, document, variable_values=None, timeout=None):
+    def execute(self, document, variable_values=None, timeout=None, url=None, headers={}):
```
Comment (lincolnq) [inline gql/transport/requests.py]: url should not be a param. headers is ok, but make it a separate pr

File: gql/transport/requests.py [hunk 3]
```diff
@@ -29,13 +29,13 @@ def execute(self, document, variable_values=None, timeout=None):
 
         data_key = 'json' if self.use_json else 'data'
         post_args = {
-            'headers': self.headers,
+            'headers': headers or self.headers,
```
Comment (lincolnq) [inline gql/transport/requests.py]: merge the dicts (use argument to override self.headers)

File: gql/response_parser.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    """The challenge is to substitute custom scalars in a GQL response with their
+    decoded counterparts.
```
Comment (lincolnq) [inline gql/response_parser.py]: This should be written more clearly: explain briefly what "custom scalars" are, and what "decoded" means in this context.

File: gql/response_parser.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    To solve this problem, we first need to iterate over all the fields in the
+    response (which is done in the `_traverse()` function).
+
+    Each time we find a field which has type scalar and is a custom scalar, we
+    need to replace the value of that field with the decoded value. All of this
+    logic happens in `_substitute()`.
+
+    Public Interface:
+    parse(): call parse with a GQL response to replace all instances of custom
+        scalar strings with their deserialized representation."""
```
Comment (lincolnq) [inline gql/response_parser.py]: remove "call parse with"; we know it's a function :) (also rename the function)

File: tests/test_response_parser.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
+"""Tests for the GraphQL Response Parser.
+
+These tests are worthless until I have a schema I can work with.
```
Comment (lincolnq) [inline tests/test_response_parser.py]: check in the starwars schema (augmented if necessary with some custom type definitions) and use it for testing, rather than depending on Money or whatever.

**Review by lincolnq — CHANGES_REQUESTED** [review]
good progress, we spoke in person

File: gql/client.py [hunk 7]
```diff
@@ -4,6 +4,8 @@
 from graphql.validation import validate
 
 from .transport.local_schema import LocalSchemaTransport
+from .type_adaptor import TypeAdaptor
```
Comment (benkuhn) [inline gql/client.py]: Most Python projects spell it `adapter`. For consistency it's probably best to spell it that way ourselves too.

File: gql/client.py [hunk 8]
```diff
@@ -17,7 +19,7 @@ def __init__(self, retries_count, last_exception):
 
 class Client(object):
     def __init__(self, schema=None, introspection=None, type_def=None, transport=None,
-                 fetch_schema_from_transport=False, retries=0):
+                 fetch_schema_from_transport=False, custom_scalars={}, retries=0):
```
Comment (benkuhn) [inline gql/client.py]: [nitpick] If we want to get this merged upstream, it's best to put new arguments at the end to preserve backwards compatibility of the function signature.

File: gql/type_adaptor.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        By default the root level is `schema.query`, if that fails, then we check
+        `schema.mutation`."""
+
+        def iterate(node: Any, lookup: List[str]):
+            lookup = lookup.copy()
+            if not lookup:
+                return self._get_scalar_type_name(node)
+
+            final_node = self._follow_type_chain(node)
+            return iterate(final_node.fields[lookup.pop(0)], lookup)
```
Comment (benkuhn) [inline gql/type_adaptor.py]: Instead of mutation, it would be more idiomatic to use slicing here (and not copy above): `return iterate(final_node.fields[lookup[0]], lookup[1:])`--the way it's currently written, it's not immediately clear whether the second argument includes the original first key or not (you have to think through when the `pop` happens).

File: gql/type_adaptor.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        this function returns the name of that type. (e.g. 'DateTime')
+
+        If it is not a scalar type (e..g a GraphQLObject or list), then this
+        function returns None.
+
+        `keys` is a breadcrumb trail telling us where to look in the GraphQL schema.
+        By default the root level is `schema.query`, if that fails, then we check
+        `schema.mutation`."""
+
+        def iterate(node: Any, lookup: List[str]):
```
Comment (benkuhn) [inline gql/type_adaptor.py]: Can you give this a better name than `iterate`?

File: gql/type_adaptor.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            lookup = lookup.copy()
+            if not lookup:
+                return self._get_scalar_type_name(node)
+
+            final_node = self._follow_type_chain(node)
+            return iterate(final_node.fields[lookup.pop(0)], lookup)
+
+        try:
```
Comment (benkuhn) [inline gql/type_adaptor.py]: This `try/except` is icky (and could theoretically cause bugs if you have a query and a mutation with the same name, although that would be stupid). What if instead this function took an argument indicating whether the root type should be `query` or `mutation`?

File: gql/type_adaptor.py [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+            return iterate(self.schema.get_query_type(), keys)
+        except (KeyError, AttributeError):
+            try:
+                return iterate(self.schema.get_mutation_type(), keys)
+            except (KeyError, AttributeError):
+                return None
+
+    def _substitute(self, keys: List[str], value: Any) -> Any:
```
Comment (benkuhn) [inline gql/type_adaptor.py]: This name makes it sound like the method mutates something. Maybe `_get_decoded_scalar_type` or something instead?

File: gql/type_adaptor.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        function on all leaf nodes. The function is called with 2 arguments:
+            keys: List[str] is a breadcrumb trail telling us where we are in the
+                response, and therefore, where to look in the GQL Schema.
+            value: Any is the value at that node in the response
+
+        Builds a new tree with the substituted values so old `response` is not
+        modified."""
+        def iterate(node: Any, keys: List[str] = []):
+            if isinstance(node, dict):
+                result = {}
```
Comment (benkuhn) [inline gql/type_adaptor.py]: You can use a dict comprehension: `return {_key: iterate(value, keys + [_key]) for key, value in node.items()}`

File: tests/test_type_adaptor.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+import requests
+from gql import Client
+from gql.transport.requests import RequestsHTTPTransport
+
+class Capitalize():
+    @classmethod
+    def parse_value(self, value: str):
+        return value.upper();
+
+@pytest.fixture
```
Comment (benkuhn) [inline tests/test_type_adaptor.py]: This makes the tests slower, dependent on the network, more likely to flake (if the network fails), and liable to break if the response at this URL changes.

Unless the project already depends on this particular URL, I'd suggest storing this in a file somewhere locally and parsing the schema yourself (the same way waveussd does it) instead of instantiating a client. In fact, even if this project already uses the URL elsewhere, it would probably be a good code improvement to remove that dependency :)

Either way, you should use `pytest.fixture(scope='session')` to prevent the fixture from being re-fetched on every request.

File: gql/type_adaptor.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+    Each time we find a field which is a custom scalar (it's type name appears
+    as a key in self.custom_scalars), we replace the value of that field with the
+    decoded value. All of this logic happens in `_substitute()`.
+
+    Public Interface:
+    apply(): pass in a GQL response to replace all instances of custom
+        scalar strings with their deserialized representation."""
+
+    def __init__(self, schema: GraphQLSchema, custom_scalars: Dict[str, Any] = {}) -> None:
```
Comment (benkuhn) [inline gql/type_adaptor.py]: Unfortunately, this project is supposed to work with Python 2.7 which doesn't support syntactic type annotations. I think it's probably worth it since this is eminently mergeable and then we could stop maintaining a fork.

File: gql/client.py (hunk 8 above)
Comment (benkuhn) [reply gql/client.py]: Also, `custom_scalars` makes it sound like the dict values are scalars, rather than adapter functions. Maybe `custom_types` or something instead?

File: gql/transport/requests.py [hunk 16]
```diff
@@ -20,16 +20,19 @@ def __init__(self, url, auth=None, use_json=False, timeout=None, **kwargs):
         self.default_timeout = timeout
         self.use_json = use_json
 
-    def execute(self, document, variable_values=None, timeout=None):
+    def execute(self, document, variable_values=None, timeout=None, headers={}):
```
Comment (benkuhn) [inline gql/transport/requests.py]: You should remove the custom headers change from this PR since it's in #2 as well. You can create a local branch with both PR's merged if you want something to run against locally (I think that's the best way to accomplish that, at least).

File: gql/type_adaptor.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+            return node.name
+        return None
+
+    def _lookup_scalar_type(self, keys: List[str]) -> Optional[str]:
+        """Search through the GQL schema and return the type identified by 'keys'.
+
+        If keys (e.g. ['film', 'release_date']) points to a scalar type, then
+        this function returns the name of that type. (e.g. 'DateTime')
+
+        If it is not a scalar type (e..g a GraphQLObject or list), then this
```
Comment (benkuhn) [inline gql/type_adaptor.py]: Isn't it misleading to say the type could be a list (because we would just look at the list's `of_type` to see whether it's a list of scalars or objects)?

File: gql/type_adaptor.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+    def __init__(self, schema: GraphQLSchema, custom_scalars: Dict[str, Any] = {}) -> None:
+        """ schema: a graphQL schema in the GraphQLSchema format
+            custom_scalars: a Dict[str, Any],
+                where str is the name of the custom scalar type, and
+                      Any is a class which has a `parse_value()` function"""
+        self.schema = schema
+        self.custom_scalars = custom_scalars
+
+    def _follow_type_chain(self, node: Any) -> Any:
```
Comment (benkuhn) [inline gql/type_adaptor.py]: This and `_get_scalar_type_name` could be pulled out of the class methods since they don't use `self`.

File: gql/type_adaptor.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+        If the type is not a custom scalar, then return the original value.
+
+        If it is a custom scalar, return the deserialized value, as
+        output by `<CustomScalarType>.parse_value()`"""
+        scalar_type = self._lookup_scalar_type(keys)
+        if scalar_type and scalar_type in self.custom_scalars:
+            return self.custom_scalars[scalar_type].parse_value(value)
+        return value
+
+    def _traverse(self, response: Dict[str, Any], substitute: Callable) -> Dict[str, Any]:
```
Comment (benkuhn) [inline gql/type_adaptor.py]: The code would be easier to read if `substitute` was hardcoded rather than an argument (since it's only ever equal to `self._substitute`) and if the name were less generic than `_traverse` (`_translate_all_scalars`, perhaps?)

Comment (benkuhn) [reply gql/type_adaptor.py]: Actually, then `apply` (which also has an excessively generic name) would be a simple wrapper and you could remove it + make this the public function

File: tests/test_type_adaptor.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    expected = copy.deepcopy(response)
+    expected['allFilms']['edges'][0]['node']['releaseDate'] = "SOME_DATETIME"
+    expected['allFilms']['edges'][1]['node']['releaseDate'] = "SOME_OTHER_DATETIME"
+
+    result = type_adaptor.apply(response)
+
+    assert result == expected
+    expected['allFilms']['edges'][0]['node']['releaseDate'] = "some_datetime"
```
Comment (benkuhn) [inline tests/test_type_adaptor.py]: What's the purpose of these two lines given that they're not followed by any assertions? Are they some kind of typo?

**Review by benkuhn — APPROVED** [review]

File: gql/client.py (hunk 8 above)
Comment (benkuhn) [reply gql/client.py]: PS: Also worth adding a docstring here that specifies the type of `custom_scalars`. (Annoying but otherwise it will be very hard to figure out!)

File: gql/type_adaptor.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+    Public Interface:
+    apply(): pass in a GQL response to replace all instances of custom
+        scalar strings with their deserialized representation."""
+
+    def __init__(self, schema: GraphQLSchema, custom_scalars: Dict[str, Any] = {}) -> None:
+        """ schema: a graphQL schema in the GraphQLSchema format
+            custom_scalars: a Dict[str, Any],
+                where str is the name of the custom scalar type, and
+                      Any is a class which has a `parse_value()` function"""
```
Comment (benkuhn) [inline gql/type_adaptor.py]: What's the signature of `parse_value`? It may be worth creating an abstract class that we expect the values in `custom_scalars` to subclass.

File: gql/type_adaptor.py (hunk 9 above)
Comment (ahussain3) [reply gql/type_adaptor.py]: good point, thanks Ben

File: gql/type_adaptor.py (hunk 11 above)
Comment (ahussain3) [reply gql/type_adaptor.py]: I think it's also a little awkward to pass in the root type because there is no natural way to determine whether it is `query` or `mutation` (and I think it provides for a cleaner interface if this function is responsible for determining that). I can refactor to make it so that we explicitly work out if it is `query` or `mutation` rather than relying on an exception to be thrown.

File: tests/test_type_adaptor.py (hunk 14 above)
Comment (ahussain3) [reply tests/test_type_adaptor.py]: Yes, local schema would be best.

The project already depends on this URL (the same pattern is used elsewhere to get the schema), and we found that because of CSRF tokens it was quite cumbersome to get the schema via the standard tools (I tried the graphql-cli, and a json-to-graphql converter). So I left it like this to avoid wasting more time.

File: gql/type_adaptor.py (hunk 17 above)
Comment (ahussain3) [reply gql/type_adaptor.py]: hmm, good point

File: gql/type_adaptor.py (hunk 18 above)
Comment (ahussain3) [reply gql/type_adaptor.py]: Is this always considered preferable? To me the functions seem relatively private, and I wouldn't imagine them needing to be called from outside of the class so I'm happy to leave them inside.

File: tests/test_type_adaptor.py (hunk 20 above)
Comment (ahussain3) [reply tests/test_type_adaptor.py]: Mistake - will fix

File: gql/type_adaptor.py (hunk 21 above)
Comment (ahussain3) [reply gql/type_adaptor.py]: At least in our case that would be a little awkward since (for example) we already have a fleshed out CurrencyAmount class. The class would inherit from multiple base classes and CurrencyAmount would be tightly coupled to the gql client, which seems like it is more trouble than it is worth.

I will update the comment to make it clear what the signature of `parse_value` is

File: gql/transport/requests.py (hunk 16 above)
Comment (ahussain3) [reply gql/transport/requests.py]: Will drop that commit from this branch 👍
